addons/membership/models/invoice.py

```diff
@@ -0,0 +1,87 @@
+# -*- coding: utf-8 -*-
+from datetime import date
+from openerp import api, models, fields, _
+import openerp.addons.decimal_precision as dp
+from openerp.exceptions import ValidationError
+
+
+class Invoice(models.Model):
+    _inherit = 'account.invoice'
+   
+    @api.multi
+    def action_cancel(self):
+        '''Create a 'date_cancel' on the membership_line object'''
+        MembershipLine = self.env['membership.membership_line']
+        today_date = date.today()
+        for invoice in self:
+            mlines = MembershipLine.search([('account_invoice_line_id', 'in',
+                                            [l.id for l in invoice.invoice_line_ids])])
+            mlines.write({'date_cancel': today_date})
+        return super(Invoice, self).action_cancel()
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+   
+    @api.multi
+    def write(self, vals):
+        """Overrides orm write method
+        """
+        MembershipLine = self.env['membership.membership_line']
+        res = super(AccountInvoiceLine, self).write(vals)
+        for line in self:
+            if line.invoice_id.type == 'out_invoice':
+                ml_ids = MembershipLine.search([('account_invoice_line_id', '=', line.id)])
+                if line.product_id and line.product_id.membership and not ml_ids:
+                    # Product line has changed to a membership product
+                    date_from = line.product_id.membership_date_from
+                    date_to = line.product_id.membership_date_to
+                    if line.invoice_id.date_invoice > date_from and line.invoice_id.date_invoice < date_to:
+                        date_from = line.invoice_id.date_invoice
+                    MembershipLine.create({
+                                    'partner_id': line.invoice_id.partner_id.id,
+                                    'membership_id': line.product_id.id,
+                                    'member_price': line.price_unit,
+                                    'date': date.today(),
+                                    'date_from': date_from,
+                                    'date_to': date_to,
+                                    'account_invoice_line_id': line.id,
+                                    })
+                if line.product_id and not line.product_id.membership and ml_ids:
+                    # Product line has changed to a non membership product
+                    MembershipLine.unlink(ml_ids)
+        return res
+    
+    @api.multi
+    def unlink(self):
+        """Remove Membership Line Record for Account Invoice Line
+        """
+        MembershipLine = self.env['membership.membership_line']
+        for membership_line in MembershipLine.search([('account_invoice_line_id', 'in', self.ids)]):
+            membership_line.unlink()
```

Comment:
why iteration ? you can directly unlink with recordsets of membership lines
membership_lines = MembershipLine.search([('account_invoice_line_id', 'in', self.ids)])
membership_lines.unlink()